### PR TITLE
Import relationships via SAF (Simple Archive Format)

### LINKED ISSUE #2883: Creating new Relationships/Entities via SAF Import
From release plan spreadsheet
Estimate from release plan: 48 hours
Expressing interest: none

Loosely related to AIP Backup & Restore #2882 (as may be similar to that implementation).

### PR DESCRIPTION
**Pull request by tysonlt**
## References
* Loosely related to #3317 (obviates the need for REST API to import relationships)
* Fixes #2883 

## Description
Add support for a 'relationships' manifest file in the Simple Archive Format. This allows items to be linked to new or existing entities during import. Items can be linked to other items in this import set by referring to their import subfolder name.

Each line in the file contains a relationship type key and an item identifier in the following format:
 
relation.<relation_key> <handle|uuid|folderName:import_item_folder|schema.element[.qualifier]:value>
 
The input_item_folder should refer the folder name of another item in this import batch. Example:

```
relation.isAuthorOfPublication 5dace143-1238-4b4f-affb-ed559f9254bb
relation.isAuthorOfPublication 123456789/1123
relation.isOrgUnitOfPublication folderName:item_001
relation.isProjectOfPublication project.identifier.id:123
relation.isProjectOfPublication project.identifier.name:A Name with Spaces
```

During initial import, new items are stored in a map keyed by the item folder name. Once the initial import is complete, a second pass checks for a 'relationships' manifest file in each folder and creates a relationship of the specified type to the specified item.

NOTE: this is only implemented for 'add' mode at this stage.

## Instructions for Reviewers
* Create a Simple Archive Format import, with entities of various types (I used Publication, Person, and Org Unit) 
* Create appropriate relationship manifest files in each import subfolder, referencing the folder name of the item to link to, or handle/UUID of existing item.
* ~~Add a "-l" or "--relationships" flag to the import command line (disabled by default)~~ Runs automatically
 
## Checklist
I have tested this on my own import dataset, but haven't created any unit/integration tests as I was unable to find any for the import utility.

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [ ] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

**Comment by tysonlt** [comment]
Currently a bit fragile with respect to resuming imports. It has to run after the main import to know how to look up items by import folder name, and the lookup map won't be complete for resumed imports. Could use the skipItems map to help with this.

**Comment by tdonohue** [comment]
Thanks @tysonlt for getting this started.  I've linked this PR up to a placeholder ticket #2883  that we had for this exact feature (as you've discovered it doesn't exist yet).  Once you feel this is ready for review by others, I can also find a volunteer or two to give it a try on their end & provide feedback.

**Comment by tysonlt** [comment]
Updated to match MetadataImport behaviour:

* Specify relationship by key (eg. relation.isAuthorOfPublication) instead of relationship database id
* Prefix folder name with "folderName:" similar to "rowName:" in MetadataImport
* Allow item lookup by unique metadata value in relationships manifest

NOTE: Added org.dspace.app.util.RelationshipUtils with method matchRelationshipType() copied from org.dspace.app.bulkedit.MetadataImport. This method could now be changed to defer to RelationshipUtils.

File: dspace-api/src/main/java/org/dspace/app/util/RelationshipUtils.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+    /**
+    * Matches two Entity types to a Relationship Type from a set of Relationship Types.
+     *
+    * @param relTypes set of Relationship Types.
+     * @param targetType entity type of target.
+     * @param originType entity type of origin referer.
+     * @return null or matched Relationship Type.
+     */
+    public static RelationshipType matchRelationshipType(List<RelationshipType> relTypes, String targetType,
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/app/util/RelationshipUtils.java:27]: Would it be possible to add a Unit Test for this method?  This seems like something we should be able to run through a unit test rather easily.

At a bare minimum, it'd be good to better document the behavior of this method...the current JavaDocs are missing params, and they don't give any examples of what the values of `originType` and `originTypeName` should look like (along with other params).  Examples and/or Unit Tests would make this method easier to understand.

File: dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportCLITool.java [hunk 2]
```diff
@@ -94,6 +94,7 @@ public static void main(String[] argv) throws Exception {
             options.addOption("R", "resume", false,
                               "resume a failed import (add only)");
             options.addOption("q", "quiet", false, "don't display metadata");
+            options.addOption("l", "relationships", false, "process relationships manifest");
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportCLITool.java:97]: Since this option only works for "add", we should note that in this description.  It could simply be updated to `"process relationships manifest (add only)"`

**Review by tdonohue — CHANGES_REQUESTED** [review]
@tysonlt : Thanks for this contribution!  Apologies for the delay in review, but I finally had a chance to look at this today.  **I was able to successfully get this working and import a Person Entity with two Publications.**  

The manner in which Entities must be created is a tad complex in SAF, but it works.  For other reviewers, here's the directory structure I had to create for testing. _This creates two Publication Entities both linked to the same Person Entity_
* `item_dir` = root directory for all item directories
    * `item_000` = First Publication
        * `contents` (created an empty file, but you can add actual file references)
        * `dubin_core.xml` - specify main dc fields like 'dc.title', dates, etc
        * `metadata_dspace.xml` - add a "dspace.entity.type" field with value of "Publication"
        * `relationships` - add one line... `relation.isAuthorOfPublication folderName:item_001`
    * `item_001` = Person
        * `contents` (created an empty file, but you can add actual file references)
        * `dubin_core.xml` - specify main dc fields (if any)
        * `metadata_dspace.xml` - add a "dspace.entity.type" field with value of "Person"
        * `metadata_person.xml` - add `person.familyName` and `person.givenName`.
        * (No "relationships" file... as I chose to link Publications back to this Person)
    * `item_0002` = Second Publication
        * `contents` (created an empty file, but you can add actual file references)
        * `dubin_core.xml` - specify main dc fields like 'dc.title', dates, etc
        * `metadata_dspace.xml` - add a "dspace.entity.type" field with value of "Publication"
        * `relationships` - add one line... `relation.isAuthorOfPublication folderName:item_001`
        
After creating that structure, I was able to run the SAF script using "add" mode and the new "-l" (--relationships") flag like this...
```
./dspace import -a -s ~/item_dir/ -e [user-email] -c [collection-handle] -m ~/mapfile.txt
```

Overall, I approve of this PR.  I do have a few minor requests inline below.  I'd also appreciate @benbosman 's feedback on the approach you took.  The new `relationships` file approach seems reasonable to me, but I'd like a second set of eyes.

Thanks overall & hopefully we can get this moved forward quickly!

File: dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportCLITool.java (hunk 2 above)
Comment (tysonlt) [reply dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportCLITool.java:97]: Gotcha. I was also wondering whether it should be enabled by default, since it will not do anything if there is no relationships manifest?

File: dspace-api/src/main/java/org/dspace/app/util/RelationshipUtils.java (hunk 1 above)
Comment (tysonlt) [reply dspace-api/src/main/java/org/dspace/app/util/RelationshipUtils.java:27]: Hi Tim, thanks for that. I just pulled this code directly from org.dspace.app.bulkedit.MetadataImport. I created a Util class just to avoid copy/pasting - the intent was that MetadataImport be refactored to refer to this Util method as well.

I can update the javadoc for this method but I don't think I'll have time allocated this month to work on a unit test.

File: dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportCLITool.java (hunk 2 above)
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportCLITool.java]: @tysonlt: Yes, that's a reasonable option to just enable this at all times.  If the `relationships` manifest isn't there, it should do nothing.  Though it also probably should be a little less "noisy" then...it can say it's "checking for relationships", but likely should then only report on which items it finds a `relationships` manifest for (and not items it finds nothing for).   

This sort of behavior could allow us to simply remove the `-l` flag altogether.  It's just always running, but it doesn't say much if no `relationships` manifests are found anywhere.

**Comment by tdonohue** [comment]
@tysonlt : This is failing unit tests as it has minor checkstyle violations.  It looks like lines 22 and 27 in the RelationshipUtils.java class both have trailing whitespace (which is not allowed).  Once you fix those, this should pass.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@tysonlt : See my comment above. This looks great overall & I think it's nearly ready.  It just has small Checkstyle issues that need fixing so that the unit tests can pass. Thanks!

**Comment by tysonlt** [comment]
Thanks @tdonohue, I've committed those changes now. I've also updated MetadataImport.java to use RelationshipUtils.java, since that's where I got that code from.

**Review by tdonohue — COMMENTED** [review]
@tysonlt : Looks good now!  I retested this and everything is still working perfectly.  I only see one more small improvement we may want to make...I think we should verify that, when `folderName` is used, a user cannot select a folder that is outside of the import directory.  So, they couldn't do something like `../../../some/other/folder/on/filesystem`.   It doesn't look to me like it's "insecure" as-is, but likely better safe than sorry.  See my small suggestion inline below.

Beyond that, I'm +1 this change. Thanks again for your hard work on this!

File: dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportServiceImpl.java [hunk 3]
```diff
@@ -288,6 +321,270 @@ public void addItems(Context c, List<Collection> mycollections,
...
+      * @param c Context
+      * @param sourceDir The parent import source directory
+      * @throws Exception
+      */
+    protected void addRelationships(Context c, String sourceDir) throws Exception {
+
+        for (Map.Entry<String, Item> itemEntry : itemFolderMap.entrySet()) {
+
+            String folderName = itemEntry.getKey();
+            String path = sourceDir + File.separatorChar + folderName;
```
[the remark this answers is not in the document]
Comment (tysonlt) [reply dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportServiceImpl.java:336]: Hi Tim, sorry for the delay. That folder is simply the name of the import item folder gathered on line 271, and used as a map key to identify items to link to on line 306:

`itemFolderMap.put(dircontents[i], item);`

As such I'm not sure it needs to be validated, as it is not coming from any user input, but from `File.list()`. It's the same method used to identify the path of the 'contents' file (which is where I nabbed it from!)

I can add that check if you like, but I'm not sure it's necessary.

**Review by benbosman — CHANGES_REQUESTED** [review]
Thanks for this PR,
I've done a partial initial review.

A few things I noticed here:
* The solution here assumes a plain-text relationships file. If you'd export an item with relationships, you'll notice the metadata_relation.xml file will contain the current relationships. It may make more sense to use that, in order to support future updates using the SAF
* The customizations here don't support the `--test` parameter. That feature is currently broken because the solution assumes the items are really created
* The `--collection` with a UUID is broken, but I noticed you didn't change that, so it may be broken in `main`

File: dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportServiceImpl.java [hunk 4]
```diff
@@ -159,6 +172,10 @@
     protected boolean useWorkflow = false;
     protected boolean useWorkflowSendEmail = false;
     protected boolean isQuiet = false;
+    protected boolean processRelationships = false;
```
[the remark this answers is not in the document]
Comment (tysonlt) [reply dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportServiceImpl.java:175]: * The rationale for a plain text file was that the items being linked to don't exist yet, since it refers to items in the current import batch. I suppose we could put special values in metadata_relation.xml, like 'folderName:item_001' instead of the linked item UUID, and do an extra pass to update those values to real UUIDs before processing that file. I find a simple manifest file simpler though.
* I completely overlooked that! Will fix.
* Yes, it didn't work for me either. I could fix that, since I can see what the problem is.

**Comment by benbosman** [comment]
Thanks for the update @tysonlt 

I think the current `relationships` file can also work, but in order to support updates with the file, the export should eventually support the `relationships` file as well.

@tdonohue what do you think about this:
* Would it be best to work with the existing xml files, and parse it as if it's metadata (similar to the CSV solution)
* Or would it be best to use the `relationships` file, which can in a future PR also support updates if the SAF export also supports it

I think both solution can also be extended to support mixing entities and plain text in a future PR

**Comment by tdonohue** [comment]
@benbosman and @tysonlt : In my opinion, this approach of having a plain text `relationships` file for *creation* of new relationships is reasonable.  I realize it's a bit odd that we have a `metadata_relation.xml` on export (because relationships are mapped to `relation.*` fields). But, I don't see an easy way to reuse the `metadata_relation.xml` file for re-import/modification, as it'd require us to process that `metadata_relation.xml` file different from all the other `metadata_*.xml` files.

So, in my opinion, we may want to consider treating the `metadata_relation.xml` as "simple, flat metadata" (just like every other `metadata_*.xml` file, while the `relationships` file is used to add/update/modify relationships.  We'll obviously have to document that clearly though, as I do see that it could get confusing.

Does that sound reasonable to both of you, or am I missing a clear reason why the `metadata_relation.xml` file would be a better approach here?

**Comment by tysonlt** [comment]
@tdonohue @benbosman I also would prefer to keep the xml as flat, static mappings to literal values, with no pre-processing just on that file, and not the others. The relationships manifest solves the very specific problem of not knowing what to put in the rel*.xml file, as the target item doesn't have a UUID yet. In this way it is similar to the other manifest files, like the collections or handle files, in that it provides instructions on how to process the item, rather than providing metadata directly.

File: dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportServiceImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportServiceImpl.java:336]: @tysonlt : My apologies for not getting back as well.  Yes, I think you are correct that this check appears unnecessary....my mistake :)

**Comment by tdonohue** [comment]
@tysonlt : I talked with @benbosman about this today, and we both agree that your current approach of using the `relationships` text file seems good enough.  We just may want to improve upon it later by also ensuring the export process can export relationships into that same `relationships` text file, so that it's easier to export & reimport from the same SAF directory.

All that means this PR can move forward as-is...we'd just need you to finish addressing any outstanding feedback, mainly the other comments from @benbosman in his review: https://github.com/DSpace/DSpace/pull/3322#pullrequestreview-756307950  Then @benbosman and I can give it a final review & test and hopefully get this merged soon.  Thanks!

File: dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportServiceImpl.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (tysonlt) [reply dspace-api/src/main/java/org/dspace/app/itemimport/ItemImportServiceImpl.java]: Fixed test mode and also fixed processing of collection parameter

**Comment by tysonlt** [comment]
I'm fixing those checkstyle errors, but I'm occasionally seeing this exception when creating relationships:

`java.lang.IllegalArgumentException: Comparison method violates its general contract!`

It is coming from org.dspace.content.DSpaceObjectServiceImpl.update(DSpaceObjectServiceImpl.java:623).

It looks like that comparator function has a missing edge case, but I can't see it!

**Review by benbosman — COMMENTED** [review]
Thanks @tysonlt 

This is looking good to me.
I did notice that incorrect data in the `relationships` is not noticed with the `--test` parameter, but that can probably be handled in a future PR

I would want some automated test to proof it keeps working though.
This can be similar to the ITs for the CSV import with relationships

**Comment by tdonohue** [comment]
@benbosman : While I understand the desire for ITs/tests in this area, unfortunately we don't yet have test infrastructure (at all) for Simple Archive Format imports/exports. This is a definite flaw (and should be fixed), but I don't think we should require this small PR to add/create that missing infrastructure.  

So, my recommendation here would be that, if we want to see this in 7.1, we should accept it as-is & log a ticket about the lack of ITs for SAF import/export.  Otherwise, I don't see any way this code will make it into 7.1....building this infrastructure for SAF testing is unfortunately not a small task.

**Review by benbosman — APPROVED** [review]
Thanks @tdonohue for that update

That would indeed increase the scope of the task too much now, but it should be fixed at a later time

**Review by tdonohue — APPROVED** [review]
👍 Re-tested and this works well. I also tested the `-c` commandline option now also accepts a Collection's UUID (it works too).  Thanks @tysonlt for your hard work on this new feature!

**Comment by tdonohue** [comment]
@tysonlt : This is at +2, so I'm merging this for release in 7.1.  I also just copied over your basic instructions into our Documentation at: https://wiki.lyrasis.org/display/DSDOC7x/Importing+and+Exporting+Items+via+Simple+Archive+Format#ImportingandExportingItemsviaSimpleArchiveFormat-relationshipsfile

Could you double check the docs look good to you as well?  Ping me on Slack if you'd like access to edit those docs (you'd need to setup a wiki account)

**Comment by tysonlt** [comment]
Great, thanks @tdonohue!

One small change for the wiki: "Entities can be linked to other entitiess in this import set..."
